### PR TITLE
Improve courses and scripts controllers before locking down courses to audiences

### PR DESCRIPTION
**Pull request by dmcavoy**
This is pre-work to locking down courses to their audiences.

This PR reorganizes the code in the scripts and courses controllers so that adding the code for locking down to audiences is easier. 

There are two changes to behavior here. One: make sure we do not try to redirect from a family name that has no stable unit groups.   Two: Prevent going to the rollup pages for a unit_group that is associated with a PLC Course

There is also improvements to some test coverage to make what the tests are doing more clear. 

## Links

[Product Spec](https://docs.google.com/document/d/1qZpc90daxQiiqpcQUiUiGV2tQoaYEQUzOA72UZzBlVE/edit#)
[Eng Plan](https://docs.google.com/document/d/1V61xONU0-mleMEEdHWNxhvZtYIPRCB31mT_hbUXusCQ/edit#heading=h.n9isd44okbq7)

## Testing story

- Existing tests with a few improvements

### CODE REVIEW

File: dashboard/app/models/unit_group.rb
Comment (dmcavoy) [inline dashboard/app/models/unit_group.rb:727]: `latest_stable_version` is the same method and this is not being used so removing.

File: dashboard/app/controllers/courses_controller.rb
Comment (dmcavoy) [inline dashboard/app/controllers/courses_controller.rb:155]: The check for `if unit_group` is new here. This makes sure we don't get Honeybadger errors for someone trying to go to a course family name that has no stable unit groups. [Example Error](https://codedotorg.slack.com/archives/C55JZ1BPZ/p1643052013005200)

File: dashboard/app/models/unit_group.rb
Comment (davidsbailey) [reply dashboard/app/models/unit_group.rb:727]: oops, I missed that earlier. nice catch!

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [inline dashboard/app/controllers/courses_controller.rb]: in general, you shouldn't have to do `return render :forbidden unless can?`, since this is pretty much exactly what `authorize_resource` does. can you see if everything still works without these calls?

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [inline dashboard/app/controllers/courses_controller.rb]: the no_access checks here look like they were previously only being done on `show`, and are now also being done on `:vocab, :resources, :code, :standards`. what's the thinking behind adding these checks to this path? 

for `:vocab, :resources, :code, :standards`, I would suggest moving this logic into CanCan, and letting it return the default of `forbidden` instead of the special `no_access` page, unless we have a clear product reason for preferring `no_access`.

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [inline dashboard/app/controllers/courses_controller.rb:172]: would you consider extracting this section into its own method with a name that more accurately describes what it's doing? I think it is a bit more readable to be able to call this as something like `before_action :check_plc_enrollment, only: [:show]` rather than embedding it in this other method.

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [inline dashboard/app/controllers/courses_controller.rb:83]: how do we know it is safe to remove lines 74-83?

File: dashboard/app/controllers/scripts_controller.rb
Comment (davidsbailey) [inline dashboard/app/controllers/scripts_controller.rb]: again, it looks like this logic belongs in can can, at least for the routes besides `show`

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [inline dashboard/app/controllers/courses_controller.rb]: I think you may be able to remove this -- the edit route will only be accessible in levelbuilder_mode, in which case get_from_cache will return the uncached version because of these lines: https://github.com/code-dot-org/code-dot-org/blob/c663c3f1f64b6d7bdc428f59a9ad73bbcc733be8/dashboard/app/models/script.rb#L436-L437

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [reply dashboard/app/controllers/courses_controller.rb:155]: checking my understanding -- is the rest of this logic (besides `if unit_group`) moving here from somewhere else?

**Review by davidsbailey — COMMENTED** [review]
very excited to see this cleanup Dani! moving authorization checks into CanCan is the way to go. many of the comments are around moving a bit further in that direction. some of the comments are checking my own understanding, so I'd like to take another look once you've had time to respond.

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [reply dashboard/app/controllers/courses_controller.rb]: if you feel strongly about no_access... this technique could make it possible to quickly override all CanCan exceptions with the friendlier page: https://github.com/code-dot-org/code-dot-org/blob/cba900f0ae44ae195462ed22b58e8c1f3d853ea4/dashboard/app/controllers/projects_controller.rb#L489

Comment (davidsbailey) [reply dashboard/app/controllers/courses_controller.rb]: and you can use this technique to handle different actions differently: https://github.com/code-dot-org/code-dot-org/blob/16a68867300a77739dcbc94fc567c498cffa041a/dashboard/app/controllers/api/v1/sections_controller.rb#L8-L10

File: dashboard/app/controllers/courses_controller.rb
Comment (dmcavoy) [reply dashboard/app/controllers/courses_controller.rb]: Ya my goal was to remove these but the tests would not pass if I did and I could not figure out why. Probably worth revisiting maybe with another set of eyes

File: dashboard/app/controllers/courses_controller.rb
Comment (dmcavoy) [reply dashboard/app/controllers/courses_controller.rb]: I was looking to make things more consistent across the different actions which is why I set it up this way. I am planning in the follow PR to show the no_access message if you can't instruct or participate in the course. I think I'll check with Tess since the spec does not really call out what to do here

File: dashboard/app/controllers/courses_controller.rb
Comment (dmcavoy) [reply dashboard/app/controllers/courses_controller.rb:172]: yes will do

File: dashboard/app/controllers/courses_controller.rb
Comment (dmcavoy) [reply dashboard/app/controllers/courses_controller.rb:83]: The logic covered here has moved into `get_unit_group` and `set_unit_group`.

File: dashboard/app/controllers/courses_controller.rb
Comment (dmcavoy) [reply dashboard/app/controllers/courses_controller.rb]: Great - i will update

File: dashboard/app/controllers/courses_controller.rb
Comment (dmcavoy) [reply dashboard/app/controllers/courses_controller.rb:155]: Its an updated version of this removed code that used to start on line 74 
```
    elsif UnitGroup.family_names.include?(params[:course_name])
      # csp and csd are each "course families", each containing multiple "course versions".
      # When the url of a course family is requested, redirect to a specific course version.
      redirect_query_string = request.query_string.empty? ? '' : "?#{request.query_string}"
      redirect_to_course = UnitGroup.latest_stable(params[:course_name])
      redirect_to "#{course_path(redirect_to_course)}#{redirect_query_string}"
```

as well as these lines from standards action

```
    if !@unit_group.present? && UnitGroup.family_names.include?(params[:course_name])
      redirect_to_course = UnitGroup.latest_stable(params[:course_name])
      redirect_to standards_course_path(redirect_to_course)
      return
    end
```

I thought it made it more consistent but if you feel like we should keep it locked down to only show and standards I could update this

Comment (dmcavoy) [reply dashboard/app/controllers/courses_controller.rb:155]: I will say things like this also match the way scripts_controller is set up so that the two are more similar

File: dashboard/app/controllers/scripts_controller.rb
Comment (dmcavoy) [reply dashboard/app/controllers/scripts_controller.rb]: Why would show want a different message than everything else? Just because thats keeping things behaving the same?

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [reply dashboard/app/controllers/courses_controller.rb]: If consistency is the goal, what do you think about consistently rendering no_access when cancan checks fail on any courses or scripts routes? that could be accomplished using rescue_from, and it would allow you to move this logic into cancan.

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [reply dashboard/app/controllers/courses_controller.rb]: ok... I'm not sure if it was worth the effort, but I got to the bottom of this.

short version: you can leave these checks here for now. :-)

medium version: the following seems to eliminate the need for these checks. explanation in comments:
```
# specifying instance_name and id_param tells CanCan to use the value of @unit_group set in set_unit_group
# when authorizing actions like vocab and resources. for more details, see
# https://github.com/ryanb/cancan/blob/705b5d6f0d8a78d9436eb90eda14e771309f155e/lib/cancan/controller_resource.rb#L161-L162
# https://github.com/ryanb/cancan/blob/705b5d6f0d8a78d9436eb90eda14e771309f155e/lib/cancan/controller_resource.rb#L137-L139 

authorize_resource class: 'UnitGroup', except: [:index], instance_name: 'unit_group', id_param: :course_name
```
however it changes some of the error codes (404 --> 403) and I don't want those to block this PR.

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [reply dashboard/app/controllers/courses_controller.rb]: if consistency is not something you want to tackle in this PR... then I would leave behavior of vocab, etc alone for now. in the case you could let those return 403 via cancan.

File: dashboard/app/controllers/courses_controller.rb
Comment (davidsbailey) [reply dashboard/app/controllers/courses_controller.rb:155]: I guess you didn't believe me when I said "checking my understanding" 😂  no worries, Dani, this looks fine

File: dashboard/app/models/ability.rb
Comment (davidsbailey) [inline dashboard/app/models/ability.rb:260]: nice shortcut! TIL

**Review by davidsbailey — APPROVED** [review]
thank you for doing all of this, Dani! Looks great!

**Pull request merged**
